### ISSUE TITLE
Uninitialized constant MonitorMixin::Thread

### ISSUE THREAD

**Issue by hurricup** [jruby/jruby#6790]
Provide at least:
* JRuby version (`jruby -v`): RVM: jruby-9.2.4.0: ver.9.2.4.0 (2.5.0) 8faff06 OpenJDK 64-Bit Server VM 11.0.11+9-Ubuntu-0ubuntu2.18.04 on 11.0.11+9-Ubuntu-0ubuntu2.18.04 +jit p0
* Operating system and platform: Ubuntu 18.04

Other relevant info you may wish to add:
* Installed or activated gems: default set of gems
* Application/framework version (e.g. Rails, Sinatra): no

From time to time, our CI tests failing with:
```
uninitialized constant MonitorMixin::Thread 
Did you mean?  Thread

org/jruby/RubyModule.java:3532:in `const_missing'
	/home/user/.rvm/rubies/jruby-9.2.4.0/lib/ruby/stdlib/monitor.rb:184:in `mon_enter'
	/home/user/.rvm/rubies/jruby-9.2.4.0/lib/ruby/stdlib/monitor.rb:224:in `mon_synchronize'
...
```

I'm not sure why this happen, but AFAIK it shouldn't. It should just use `Thread` (and it does most of the time) 

Basically it treats `Threads` as `MonitorMixin::Thread` despite `require 'thread'` in the beginning of `monitor.rb` which defines `Thread`

Not 100% sure but happening on other versions of JRuby as well.

**Comment by hurricup** [comment]
Same thing on 9.2.19.0

```
org/jruby/RubyModule.java:3766:in `const_missing'
	/home/user/.rbenv/versions/jruby-9.2.19.0/lib/ruby/stdlib/monitor.rb:233:in `mon_synchronize'
```

**Comment by hurricup** [comment]
Same thing in different place, `bundle install`  (jruby-9.2.19.0: ver.9.2.19.0 (2.5.8) 55810c5 OpenJDK 64-Bit Server VM 11.0.11+9-Ubuntu-0ubuntu2.18.04 on 11.0.11+9-Ubuntu-0ubuntu2.18.04 +jit p0)

```
NameError: uninitialized constant Gem::Version
stderr: Did you mean?  Gem::VERSION
stderr: org/jruby/RubyModule.java:3766:in `const_missing'
stderr: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/gems/shared/gems/bundler-2.2.25/lib/bundler/feature_flag.rb:37:in `block in FeatureFlag'
stderr: org/jruby/RubyBasicObject.java:2622:in `instance_eval'
stderr: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/gems/shared/gems/bundler-2.2.25/lib/bundler/feature_flag.rb:22:in `block in plugins?'
stderr: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/gems/shared/gems/bundler-2.2.25/lib/bundler/plugin.rb:220:in `hook'
stderr: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/gems/shared/gems/bundler-2.2.25/lib/bundler/installer/parallel_installer.rb:182:in `do_install'
stderr: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/gems/shared/gems/bundler-2.2.25/lib/bundler/installer/parallel_installer.rb:177:in `block in worker_pool'
stderr: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/gems/shared/gems/bundler-2.2.25/lib/bundler/worker.rb:62:in `apply_func'
stderr: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/gems/shared/gems/bundler-2.2.25/lib/bundler/worker.rb:57:in `block in process_queue'
stderr: org/jruby/RubyKernel.java:1442:in `loop'
stderr: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/gems/shared/gems/bundler-2.2.25/lib/bundler/worker.rb:54:in `process_queue'
stderr: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/gems/shared/gems/bundler-2.2.25/lib/bundler/worker.rb:91:in `block in create_threads'
```

**Comment by hurricup** [comment]
Instability of this issue may indicate some race condition from my experience

**Comment by hurricup** [comment]
Same on `jgem install rubocop`

```
warning: thread "Ruby-0-Thread-3: /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/request_set.rb:168" terminated with exception (report_on_exception is true):
NameError: uninitialized constant Gem::Specification
Did you mean?  Gem::StubSpecification
const_missing at org/jruby/RubyModule.java:3766
== at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/resolver/activation_request.rb:28
install at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/request_set.rb:173
ERROR:  While executing gem ... (NameError)
uninitialized constant Gem::Specification
Did you mean?  Gem::StubSpecification
```

**Comment by hurricup** [comment]
Not sure if this one is the same, but may be the same cause (unstable reproduction, pretty rare) on `jgem install rubocop`
```
warning: thread "Ruby-0-Thread-3: /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/request_set.rb:168" terminated with exception (report_on_exception is true):
NoMethodError: undefined method `ignored=' for nil:NilClass
to_spec at /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/stub_specification.rb:196
map at org/jruby/RubyArray.java:2588
_all at /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/specification.rb:744
each at /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/specification.rb:954
any? at org/jruby/RubyEnumerable.java:1872
installed? at /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/resolver/activation_request.rb:104
install at /usr/share/rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/request_set.rb:174
ERROR:  While executing gem ... (NoMethodError)
undefined method `ignored=' for nil:NilClass
```

**Comment by headius** [comment]
This could be a race involving an autoloaded constant. It may be fixed on master due to improvements in load/autoload/require logic.

```
uninitialized constant MonitorMixin::Thread 
Did you mean?  Thread
```

This is particularly odd since the top-level `Thread` constant should be available from everywhere.

The others could be the same issue or something different, but I am unclear exactly how to reproduce. I imagine it is not possible for me to run your project CI but for the other cases, did you literally just install a gem and get that `Specification` error?

These errors are strange but I very much want to help fix them if we are doing something wrong!

**Comment by hurricup** [comment]
You can't have access to our project, but there is nothing fancy going on in Specifications/Gem::Version case. Just running `bundle install` or `jgem install`. That's it. In bundler case, after few successful gem installations, some random installation fails.
I could not reproduce this bug intentionally. I will provide additional information about cases.

**Comment by hurricup** [comment]
jruby-9.2.19.0: ver.9.2.19.0 (2.5.8) 55810c552b OpenJDK 64-Bit Server VM 11.0.11+9-Ubuntu-0ubuntu2.18.04 on 11.0.11+9-Ubuntu-0ubuntu2.18.04 +jit p0 (/home/user/.rubies/jruby-9.2.19.0/bin/jruby)

```
> /home/user/.rubies/jruby-9.2.19.0/bin/jruby /home/user/.rubies/jruby-9.2.19.0/bin/jgem install simplecov --version '= 0.21.2' --no-document

warning: thread "Ruby-0-Thread-3: /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/request_set.rb:168" terminated with exception (report_on_exception is true):

stderr: NameError: uninitialized constant Gem::Platform
Did you mean?  PLATFORM
const_missing at org/jruby/RubyModule.java:3766
match at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/platform.rb:29
any? at org/jruby/RubyArray.java:4552
match at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/platform.rb:26
stubs at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/specification.rb:817
select at org/jruby/RubyArray.java:2667
stubs at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/specification.rb:817
_all at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/specification.rb:744
each at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/specification.rb:954
any? at org/jruby/RubyEnumerable.java:1872
installed? at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/resolver/activation_request.rb:104
install at /home/user/.rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/request_set.rb:174
While executing gem ... (NameError)
uninitialized constant Gem::Platform
Did you mean?  PLATFORM
```

**Comment by hurricup** [comment]
One more. This one different, but probably the same cause:
```
> /home/user/.rvm/bin/rvm jruby-9.2.19.0 do /home/user/.rvm/rubies/jruby-9.2.19.0/bin/jruby /home/user/.rvm/rubies/jruby-9.2.19.0/bin/jgem install puppet --version '= 7.10.0' --no-document
 warning: thread "Ruby-0-Thread-7: /home/user/.rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/request_set.rb:168" terminated with exception (report_on_exception is true):
 NoMethodError: undefined method `ignored=' for nil:NilClass
 to_spec at /home/user/.rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/stub_specification.rb:196
 map at org/jruby/RubyArray.java:2588
 _all at /home/user/.rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/specification.rb:744
 each at /home/user/.rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/specification.rb:954
 any? at org/jruby/RubyEnumerable.java:1872
 installed? at /home/user/.rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/resolver/activation_request.rb:104
 install at /home/user/.rvm/rubies/jruby-9.2.19.0/lib/ruby/stdlib/rubygems/request_set.rb:174
 ERROR:  While executing gem ... (NoMethodError)
 undefined method `ignored=' for nil:NilClass
Process terminated: 1
```

**Comment by headius** [comment]
Sorry I have not had time to circle back to this issue!

Your errors are very peculiar and I would initially guess something is wrong with our installation. None of the commands you have shown should be doing anything exotic with concurrency.

Can you show me what `gem -v` prints? Perhaps rbenv is installing a different RubyGems version that has an issue?

**Comment by hurricup** [comment]
Here are docker images we are running our tests in:
```
docker pull rubymine/test_environment:rvm-single-ssh-jruby-92
docker pull rubymine/test_environment:rvm-single-ssh-jruby-91
docker pull rubymine/test_environment:rbenv-ssh-user
```
Currently we have only last versions.
There are also different version managers, but errors seems happening everywhere.

**Comment by hurricup** [comment]
Regarding the concurrency, I was thinking more about some jruby internal parallelization, unrelated to specific gems or code.  Haven't looked at jruby code yet.

**Comment by hurricup** [comment]
Jruby 9.3.0 update:

Looks like problem with constant resolve is gone (so far at least).  Still `Nil` appeared few times in unexpected places. Like:

```
$ bash -c "/home/user/.rvm/bin/rvm jruby-9.3.0.0 do /home/user/.rvm/rubies/jruby-9.3.0.0/bin/jruby /home/user/.rvm/rubies/jruby-9.3.0.0/bin/jgem install steep --version '= 0.46.0' --no-document
ERROR:  While executing gem ... (NoMethodError)
undefined method `name' for nil:NilClass
```
No stacktrace for some reason.

**Comment by headius** [comment]
Hmm ok. I'm not convinced that it is fixed since you are still getting some odd errors. Still not easy to reproduce I assume?

**Comment by hurricup** [comment]
Nope.

**Comment by hurricup** [comment]
Added debugging options to our CI few days ago. So far - no failures :D
Let's wait.

**Comment by hurricup** [comment]
Got a gem install failure with log and debug options:
```
$ bash -c ". /usr/local/share/chruby/chruby.sh && chruby jruby-9.3.2.0 && /home/user/.rubies/jruby-9.3.2.0/bin/jruby -Xlog.exceptions /home/user/.rubies/jruby-9.3.2.0/bin/jgem install pry --version '= 0.13.1' --no-document --backtrace --verbose --debug"
 2022-01-04T04:27:25.230+03:00 [main] INFO TraceType : Exception raised: NameError : uninitialized constant Date
 Did you mean?  Data
 2022-01-04T04:27:25.297+03:00 [main] INFO TraceType : Exception raised: NameError : uninitialized constant BigDecimal
 2022-01-04T04:27:25.332+03:00 [main] INFO TraceType : Exception raised: NameError : uninitialized constant DateTime
 NOTE:  Debugging mode prints all exceptions even when rescued
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::AF_INET6
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::IPPROTO_ICMPV6
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::IPPROTO_IPV6
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::PF_INET6
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/socket.rb:4: warning: already initialized constant Socket::Constants::SOL_IPV6
04:27:27,929 stdout: HEAD https://index.rubygems.org/
04:27:28,180  stdout: 200 OK
04:27:28,191  stdout: GET https://index.rubygems.org/info/pry
04:27:28,289  stdout: 200 OK
04:27:28,571  stdout: GET https://index.rubygems.org/quick/Marshal.4.8/pry-0.13.1-java.gemspec.rz
04:27:28,613  stdout: 200 OK
 2022-01-04T04:27:28.612+03:00 [main] INFO TraceType : Exception raised: Errno::EEXIST : File exists - /home/user/.local/share/gem/specs/index.rubygems.org%443/quick/Marshal.4.8
04:27:28,929  stdout: GET https://index.rubygems.org/info/coderay
04:27:28,986  stdout: 200 OK
04:27:28,992  stdout: GET https://index.rubygems.org/info/method_source
04:27:29,034  stdout: 200 OK
04:27:29,040  stdout: GET https://index.rubygems.org/info/spoon
04:27:29,087  stdout: 200 OK
04:27:29,166  stdout: GET https://index.rubygems.org/info/ffi
04:27:29,223  stdout: 200 OK
04:27:29,531  stdout: GET https://index.rubygems.org/quick/Marshal.4.8/spoon-0.0.6.gemspec.rz
04:27:29,608  stdout: 200 OK
04:27:29,608  stdout: GET https://index.rubygems.org/quick/Marshal.4.8/method_source-1.0.0.gemspec.rz
 2022-01-04T04:27:29.612+03:00 [Ruby-0-Thread-2: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167] INFO TraceType : Exception raised: Errno::EEXIST : File exists - /home/user/.local/share/gem/specs/index.rubygems.org%443/quick/Marshal.4.8
04:27:29,687  stdout: 200 OK
04:27:29,687  stdout: GET https://index.rubygems.org/quick/Marshal.4.8/coderay-1.1.3.gemspec.rz
 2022-01-04T04:27:29.687+03:00 [Ruby-0-Thread-3: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167] INFO TraceType : Exception raised: Errno::EEXIST : File exists - /home/user/.local/share/gem/specs/index.rubygems.org%443/quick/Marshal.4.8
04:27:29,732  stdout: 200 OK
 2022-01-04T04:27:29.731+03:00 [Ruby-0-Thread-5: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167] INFO TraceType : Exception raised: Errno::EEXIST : File exists - /home/user/.local/share/gem/specs/index.rubygems.org%443/quick/Marshal.4.8
04:27:29,842  stdout: Downloading gem method_source-1.0.0.gem
04:27:29,842  stdout: Downloading gem coderay-1.1.3.gem
04:27:29,843  stdout: Downloading gem pry-0.13.1-java.gem
04:27:29,844  stdout: Downloading gem spoon-0.0.6.gem
 2022-01-04T04:27:29.844+03:00 [Ruby-0-Thread-5: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167] INFO TraceType : Exception raised: Errno::ENOENT : No such file or directory - /home/user/.gem/jruby/2.6.8/cache/coderay-1.1.3.gem
 2022-01-04T04:27:29.844+03:00 [Ruby-0-Thread-4: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167] INFO TraceType : Exception raised: Errno::ENOENT : No such file or directory - /home/user/.gem/jruby/2.6.8/cache/pry-0.13.1-java.gem
 2022-01-04T04:27:29.845+03:00 [Ruby-0-Thread-3: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167] INFO TraceType : Exception raised: Errno::ENOENT : No such file or directory - /home/user/.gem/jruby/2.6.8/cache/method_source-1.0.0.gem
 2022-01-04T04:27:29.846+03:00 [Ruby-0-Thread-2: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167] INFO TraceType : Exception raised: Errno::ENOENT : No such file or directory - /home/user/.gem/jruby/2.6.8/cache/spoon-0.0.6.gem
04:27:29,850  stdout: GET https://index.rubygems.org/gems/pry-0.13.1-java.gem
 2022-01-04T04:27:29.864+03:00 [Ruby-0-Thread-1: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167] INFO TraceType : Exception raised: NoMethodError : undefined method `ignored=' for nil:NilClass
 warning: thread "Ruby-0-Thread-1: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167" terminated with exception (report_on_exception is true):
 
 NoMethodError: undefined method `ignored=' for nil:NilClass
 to_spec at /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/stub_specification.rb:193
 map at org/jruby/RubyArray.java:2642
 _all at /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/specification.rb:748
 each at /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/specification.rb:953
 any? at org/jruby/RubyEnumerable.java:1710
 installed? at /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/resolver/activation_request.rb:111
 install at /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:173
 Exception `NoMethodError' at /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/stub_specification.rb:193 - undefined method `ignored=' for nil:NilClass
 ERROR:  While executing gem ... (NoMethodError)
 undefined method `ignored=' for nil:NilClass
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/stub_specification.rb:193:in `to_spec'
 org/jruby/RubyArray.java:2642:in `map'
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/specification.rb:748:in `_all'
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/specification.rb:953:in `each'
 org/jruby/RubyEnumerable.java:1710:in `any?'
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/resolver/activation_request.rb:111:in `installed?'
 /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:173:in `block in install'
 2022-01-04T04:27:29.868+03:00 [main] INFO TraceType : Exception raised: Gem::SystemExitException : Exiting RubyGems with exit_code 1
 Exception `Gem::SystemExitException' at /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/user_interaction.rb:365 - Exiting RubyGems with exit_code 1
 2022-01-04T04:27:29.869+03:00 [main] INFO TraceType : Exception raised: SystemExit : exit
 2022-01-04T04:27:29.895+03:00 [Ruby-0-Thread-4: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167] INFO TraceType : Exception raised: IOError : closed stream
04:27:29,935  stdout: GET https://index.rubygems.org/gems/method_source-1.0.0.gem
 warning: thread "Ruby-0-Thread-4: /home/user/.rubies/jruby-9.3.2.0/lib/ruby/stdlib/rubygems/request_set.rb:167" terminated with exception (report_on_exception is true):
```

**Comment by deivid-rodriguez** [comment]
I've seen this in our CI too at our repo https://github.com/rubygems/rubygems. I don't know the root cause, but I prepared https://github.com/rubygems/rubygems/pull/5228 that should fix this particular error.

**Comment by hurricup** [comment]
updated rubygems on CI to 3.3.5 go get this update. Will see if it helps.

**Comment by deivid-rodriguez** [comment]
🤞
